src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,155 @@
 package se.liu.ida.hefquin.engine.datastructures.impl;
 
-import java.util.Iterator;
+import java.util.*;
 
 import org.apache.jena.graph.Node;
 import org.apache.jena.sparql.core.Var;
 
+import org.apache.jena.sparql.engine.binding.Binding;
 import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
 
 public class SolutionMappingsHashTable extends SolutionMappingsIndexBase
 {
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Set<Var> joinVariables;
+
+	public SolutionMappingsHashTable(final ExpectedVariables... inputVars){
+		assert inputVars.length == 2;
+		final Set<Var> certainVars = new HashSet<>( inputVars[0].getCertainVariables());
+		certainVars.retainAll( inputVars[1].getCertainVariables() );
+
+		this.joinVariables = certainVars;
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		return map.size();
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		return map.isEmpty();
 	}
 
 	@Override
 	public boolean contains( final Object o ) {
-		// TODO Auto-generated method stub
-		return false;
+		return map.containsKey(o);
 	}
 
 	@Override
 	public Iterator<SolutionMapping> iterator() {
-		// TODO Auto-generated method stub
-		return null;
+		final List<SolutionMapping> solMap = new ArrayList<>();
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while(li.hasNext()){
+			solMap.addAll(li.next());
+		}
+		return solMap.iterator();
 	}
 
 	@Override
 	public boolean add( final SolutionMapping e ) {
-		// TODO Auto-generated method stub
-		return false;
+		final Binding solMapBinding = e.asJenaBinding();
+		final List<Node> valKeyL = new ArrayList<>();
+
+		for (Var v : joinVariables) {
+			valKeyL.add(solMapBinding.get(v));
```

Comment:
There should be a check that the given solution mappings is indeed defined for all of these variables. Throw an illegal argument exception if it is not. This would be mainly a safety measure in case there are programming errors in the code that is using this data structure.